### PR TITLE
[ADD] sale_packaging_price: New module to assign a package price.

### PR DESCRIPTION
**Pull request by sergio-teruel**
This module was written to extends the functionality of sale workflow and allow you to assign a selling price of a complete package.
This price is split for having the price unit according the number of pieces that fits in that packaging, so there can be rounding issues to get the exact price that are warned if happens.

### CODE REVIEW

**Comment by rafaelbn** [comment]
I cannot test it, something wrong in runbot. Please could you review? Thanks

File: sale_packaging_price/README.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+This module was written to extends the functionality of sale workflow
+and allow you to assign a selling price of a complete package.
+
+Usage
+=====
+
+To use this module, you need to:
+
+* go to warehause config and check 'Use packages: pallets, boxes, ...' and
```
Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: s/warehause/warehouse

Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: This should be on Configuration section

File: sale_packaging_price/README.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/167/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/167/issues/new?body=module:%20sale_packaging_price%0Aversion:%208.0.1.0.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
+
+License
```
Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: Why this additional section? It's already specified on the flag.

File: sale_packaging_price/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# (c) 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    'name': "sale_packaging_price",
```
Comment (pedrobaeza) [inline sale_packaging_price/__openerp__.py]: Put a real title here

**Comment by pedrobaeza** [comment]
Remove sale_packaging_price/i18n/.gitkeep

File: sale_packaging_price/views/sale_packaging_price_menu.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data noupdate="1">
+	<menuitem id="menu_product_packaging_action"
+			  action="action_product_packaging_form"
+              parent="base.menu_product" sequence="99" />
+</data>
+</openerp>
```
Comment (pedrobaeza) [inline sale_packaging_price/views/sale_packaging_price_menu.xml]: Remove extra spaces and put a line feed

File: sale_packaging_price/views/sale_stock_view.xml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            </xpath>
+           <xpath expr="//field[@name='order_line']/form/group/group/field[@name='price_unit']"
+                  position="after">
+               <field name="packaging_price" string="Package Price"/>
+           </xpath>
+        </field>
+    </record>
+
+</data>
+</openerp>
```
Comment (pedrobaeza) [inline sale_packaging_price/views/sale_stock_view.xml]: Line feed

File: sale_packaging_price/views/product_view.xml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        <field name="view_mode">tree,form</field>
+        <field name="help" type="html">
+          <p class="oe_view_nocontent_create">
+            Click to add a packaging for your products.
+          </p>
+        </field>
+    </record>
+
+</data>
+</openerp>
```
Comment (pedrobaeza) [inline sale_packaging_price/views/product_view.xml]: Line feed

**Comment by pedrobaeza** [comment]
You need to add _product_packaging_ field to the form view.

File: sale_packaging_price/models/sale_stock.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
```
Comment (pedrobaeza) [inline sale_packaging_price/models/sale_stock.py]: Why name it sale_stock.py instead sale_order.py?

File: sale_packaging_price/models/sale_stock.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+from openerp import models, fields, api
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    packaging_price = fields.Float(
+        related='product_packaging.list_price', string='Package Price',
+        readonly=True)
+    product_packaging = fields.Many2one(ondelete='restrict')
```
Comment (pedrobaeza) [inline sale_packaging_price/models/sale_stock.py]: You have to add a domain to restrict package selection to those that the product allows.

**Comment by pedrobaeza** [comment]
You have to also intercept _product_id_change_ for handling changes on quantity. You should also prevent when packaging is selected, non multiple quantities.

File: sale_packaging_price/models/sale_stock.py (hunk 7 above)
Comment (sergio-teruel) [inline sale_packaging_price/models/sale_stock.py]: I rewrite function _product_packaging_change_ that is located into sale_stock module in sale_stock.py file

File: sale_packaging_price/models/sale_stock.py (hunk 8 above)
Comment (sergio-teruel) [inline sale_packaging_price/models/sale_stock.py]: That domain allready exists in sale_stock module. Only rewrite product_packaging field for prevent delete packages that have been sold

**Comment by sergio-teruel** [comment]
@rafaelbn @pedrobaeza Changes made

File: sale_packaging_price/README.rst [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+Configuration
+=============
+
+To use this module, you need to:
+
+* go to warehouse config and check 'Use packages: pallets, boxes, ...' and
+'Allow to define several packaging methods on products'
+
+* go to product and into procurement tab you can add new packages with a
```
Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: This part is Usage section. I was talking about the other paragraph

File: sale_packaging_price/README.rst [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+======================
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/167/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/167/issues/new?body=module:%20sale_packaging_price%0Aversion:%208.0.1.0.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: Odoo version, 8.0 (not module version)

File: sale_packaging_price/views/sale_packaging_price_menu.xml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data noupdate="1">
+
+	<menuitem id="menu_product_packaging_action"
+			  action="action_product_packaging_form"
+			  parent="base.menu_product" sequence="99" />
```
Comment (pedrobaeza) [inline sale_packaging_price/views/sale_packaging_price_menu.xml]: This should have the security group for packaging

**Comment by sergio-teruel** [comment]
@rafaelbn @pedrobaeza Chnges made

**Comment by pedrobaeza** [comment]
:+1:

File: sale_packaging_price/README.rst [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+* https://www.odoo.com/forum/help-1
+
+Known issues / Roadmap
+======================
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/167/issues>`_.
```
Comment (yajo) [inline sale_packaging_price/README.rst]: 167?

File: sale_packaging_price/models/sale_stock.py (hunk 7 above)
Comment (yajo) [inline sale_packaging_price/models/sale_stock.py]: Upstream is wrong then: https://www.odoo.com/documentation/8.0/reference/guidelines.html

File: sale_packaging_price/report/sale_report.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# (c) 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from openerp import models, fields
+
+
+class SaleReport(models.Model):
```
Comment (yajo) [inline sale_packaging_price/report/sale_report.py:9]: Shouldn't this file go into `/models`?

File: sale_packaging_price/views/product_view.xml [hunk 14]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+<data>
+
+    <record id="sale_packaging_price_product_tree_inherit" model="ir.ui.view">
```
Comment (yajo) [inline sale_packaging_price/views/product_view.xml]: This is not a requirement, but just letting you know that ids get automatic prefix by the module name, so this xmlid will become `sale_packaging_price.sale_packaging_price_product_tree_inherit`, which is redundant as you see.

**Comment by yajo** [comment]
Testing in runbot, I created a package with 6 wine bottles (individual price would be 20€, pack price 100€).

See the results:
captura de pantalla de 2015-10-20 17-06-36

100.02€ is not the price I set.

To avoid this kind of problems, add a constraint that will not let the user to set an icompatible unit + package price combination. It will hassle them, but better hassle the user than the client.

File: sale_packaging_price/README.rst [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+* go to warehouse config and check 'Use packages: pallets, boxes, ...' and
+'Allow to define several packaging methods on products'
+
+Usage
+=====
+
+To use this module, you need to:
+
+* go to product and into procurement tab you can add new packages with a
```
Comment (yajo) [inline sale_packaging_price/README.rst]: It's in the inventory tab.

File: sale_packaging_price/README.rst [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+sale price
```
Comment (yajo) [inline sale_packaging_price/README.rst]: 2 spaces before here.

File: sale_packaging_price/README.rst [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+* This module also adds a new entry into product menu called 'Packaging' to
+add quickly combinations packages
```
Comment (yajo) [inline sale_packaging_price/README.rst]: 2 spaces before here too.

File: sale_packaging_price/README.rst [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+This module was written to extends the functionality of sale workflow
+and allow you to assign a selling price of a complete package.
+
+Configuration
+=============
+
+To configure this module, you need to:
+
+* go to warehouse config and check 'Use packages: pallets, boxes, ...' and
+'Allow to define several packaging methods on products'
```
Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: 2 spaces

File: sale_packaging_price/models/sale_stock.py (hunk 7 above)
Comment (pedrobaeza) [inline sale_packaging_price/models/sale_stock.py]: No, it's not wrong. You should rename it to sale_order (model name, not module name).

File: sale_packaging_price/models/sale_stock.py (hunk 8 above)
Comment (pedrobaeza) [inline sale_packaging_price/models/sale_stock.py]: OK

File: sale_packaging_price/report/sale_report.py (hunk 13 above)
Comment (sergio-teruel) [inline sale_packaging_price/report/sale_report.py:9]: I'm not sure, but i have seen that reports ('analysis') folders contains all files related to report.
@pedrobaeza can you help us about this??

File: sale_packaging_price/views/product_view.xml (hunk 14 above)
Comment (sergio-teruel) [inline sale_packaging_price/views/product_view.xml]: :ok:

File: sale_packaging_price/report/sale_report.py (hunk 13 above)
Comment (pedrobaeza) [inline sale_packaging_price/report/sale_report.py:9]: I see this correct

File: sale_packaging_price/README.rst (hunk 15 above)
Comment (sergio-teruel) [inline sale_packaging_price/README.rst]: Ok.
About the previous comment(100.02) you can increment the price decimal precision.

**Comment by sergio-teruel** [comment]
@Yajo @pedrobaeza @rafaelbn Changes made.

File: sale_packaging_price/models/sale_stock.py (hunk 7 above)
Comment (yajo) [inline sale_packaging_price/models/sale_stock.py]: Quoting from there:

> File naming: [...]
> models/< inherited_main_model >.py

And class name is correct in CamelCase, also as guidelines say.

File: sale_packaging_price/report/sale_report.py (hunk 13 above)
Comment (yajo) [inline sale_packaging_price/report/sale_report.py:9]: OK, just asking.

File: sale_packaging_price/models/sale_stock.py (hunk 7 above)
Comment (pedrobaeza) [inline sale_packaging_price/models/sale_stock.py]: I still don't know what's wrong with my proposal: rename it to sale_order.py

**Comment by yajo** [comment]
Much better, but still getting https://github.com/OCA/sale-workflow/pull/221#issuecomment-149596550 though. Fix it and test it please.

File: sale_packaging_price/models/sale_stock.py (hunk 7 above)
Comment (yajo) [inline sale_packaging_price/models/sale_stock.py]: Maybe it was a mistake, you said:

> You should rename it to sale_order (**model name**, not module name).

The model name is `SaleOrderLine`, correct.

Indeed the **module** should be `sale_order.py`.

Comment (pedrobaeza) [inline sale_packaging_price/models/sale_stock.py]: No, the main model is sale.order, not sale.order.line, that's why I ask for sale_order.py

Comment (yajo) [inline sale_packaging_price/models/sale_stock.py]: Ah OK, I was understanding that you wanted him to rename the model! We agree then :wink:

**Comment by sergio-teruel** [comment]
@Yajo about  https://github.com/OCA/sale-workflow/pull/221#issuecomment-149596550 you can increment the price decimal precision.

**Comment by yajo** [comment]
Not sure how to do that. Can you explain?

Anyway, independently of the decimal precision, if you set a final price for a bundle of products, that should always be the final price. If the final price cannot be exactly matched because of the mix of quantity, price and decimal precision, then it should not allow to put that price in the first place.

**Comment by carlosdauden** [comment]
cc @pedrobaeza @Yajo @rafaelbn @sergio-incaser

File: sale_packaging_price/README.rst [hunk 19]
```diff
@@ -0,0 +1,75 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+====================
+Sale Packaging Price
+====================
+
+This module was written to extends the functionality of sale workflow
+and allow you to assign a selling price of a complete package.
```
Comment (pedrobaeza) [inline sale_packaging_price/README.rst]: Include this text: "This price is split for having the price unit according the number of pieces that fits in that packaging, so there can be rounding issues to get the exact price that are warned if happens."

**Comment by yajo** [comment]
The code looks good, but there is a new bug.
1. Add a new product.
2. Go to Inventory > Packaging > Configurations.
3. Add new.

Error:

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/addons/web/controllers/main.py", line 944, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/api.py", line 372, in old_api
    result = method(recs, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/models.py", line 5929, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/models.py", line 5787, in _onchange_eval
    method_res = method(self)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3119204-221-bd055f/openerp/addons/sale_packaging_price/models/product.py", line 23, in _onchange_list_price
    round(self.list_price / self.qty, price_precision) *
ZeroDivisionError: float division by zero
```

**Comment by yajo** [comment]
Seems to happen when qty = 0. When saving a wrong package price, the warning works as expected.

**Comment by carlosdauden** [comment]
@pedrobaeza @Yajo @rafaelbn Changes made

File: sale_packaging_price/README.rst (hunk 19 above)
Comment (carlos-incaser) [inline sale_packaging_price/README.rst]: :+1: Added.

**Comment by carlosdauden** [comment]
@pedrobaeza Changes made

**Comment by carlosdauden** [comment]
cc @Yajo @rafaelbn

**Comment by rafaelbn** [comment]
About this: https://github.com/OCA/sale-workflow/pull/221#issuecomment-151565812 I agree with @Yajo . Check: Setting -> Technical -> Database structure -> Decimal accuracy

**Comment by rafaelbn** [comment]
To solve problem of decimal precision please add in usage (even-though this is not the solution):
- Go to Setting -> Technical -> Database structure -> Decimal accuracy, select  Product Price and set Digits to 3, if you get better accuracy set 4 digits of higher.

Been exactly: (maybe I need some illustration here)

> sale_packaging_price: New module to assign a package price
- A package in Odoo is object stock.quant.package - Go To Warehouse -> Products -> Packages.
- A Logistic Unit in Odoo is object product.ul - Go To Warehouse -> Configuration -> Products -> Logistic Units. Also here I cannot see price.
- A product packaging in Odoo is object product.packaging - Go To Warehouse -> Products -> Products packaging. In the product form (tab Inventory) this is called Packaging configuration.

Having this clear (I'm not sure if I have) the objective of module:
- To sell 1 product in packages, then if you make a quotation or invoice you are not selling 6 units of individual product, you are selling 1 unit of the package. Thanks why you set the price in the package and not in the product.
  - With this point of view you can guess that the problem of decimal precision disappears as if you say a package/box cost 100 and you sell 1 unit of a package.
  - So you buy individual units or packages/boxes but you store units
    For example:
  - You buy 1 package of 12 bottles of cola (6€) and you stock is 12 bottles of cola, each one cost 0.50€
  - You sell packages of 2 bottles of cola, each package has sell price 2€.

**Comment by rafaelbn** [comment]
@

**Comment by rafaelbn** [comment]
What do you think about this? https://github.com/OCA/sale-workflow/pull/221#issuecomment-156627097
cc @liebana

**Comment by rafaelbn** [comment]
@sergio-incaser runbot is failing, could you please check it? thanks

**Comment by liebana** [comment]
We have something for one customer, we based in the old product-pack of v6.1. You first define a service product as a bundle:

seleccion_999 076

and then you select which are their components:

seleccion_999 077

Then, because the first product is a service and there can be different VATs on the components in our case, we detail everything in the sale order

image

but only the stock of the components is carried to the delivery order.

But this module is not finished, because is intended to work with Magento connector and right now it's not possible to create from the ERP the sale order itself. We can upload the code anyway if you want.

This is the way we would eventually solve your requirement, but I haven't look your solution.

**Comment by yajo** [comment]
IMHO this solution is better integrated with current core Odoo. The notice of decimal precision seems fair enough to me. If you want 1.99€/pack, you will be warned, but not disabled to do it.

**Comment by rafaelbn** [comment]
Hi @liebana could you make a PR in https://github.com/OCA/product-attribute/tree/8.0 with that functionality? Module name could be product_pack, this could be better to compare. Ping us when done please, thanks.

**Comment by yajo** [comment]
:+1: to this PR, but @liebana's implementation is interesting too. Could you send it please?

**Comment by pedrobaeza** [comment]
That module is in 8.0 in https://github.com/ingadhoc/odoo-addons/tree/8.0/product_pack, but it has the problems I mention here: https://github.com/OCA/product-attribute/pull/6#issuecomment-49025210

**Comment by rafaelbn** [comment]
Thanks! Tested in runbot :+1:

**Pull request merged**